### PR TITLE
Adds BIP84 as an allowed derivation path purpose

### PR DESCRIPTION
**Pull request by alex-miller-0**
This allows for export of bech32 addresses

### CODE REVIEW

**Review by KevinVitale — APPROVED** [review]

File: src/bitcoin.js [hunk 1]
```diff
@@ -203,7 +192,22 @@ exports.serializeTx = function(data) {
...
+  let bech32Prefix = null;
+  let bech32Version = null;
+  if (version === FMT_SEGWIT_NATIVE_V0) {
+    bech32Prefix = SEGWIT_NATIVE_V0_PREFIX;
+    bech32Version = SEGWIT_V0;
+  } else if (version === FMT_SEGWIT_NATIVE_V0_TESTNET) {
+    bech32Prefix = SEGWIT_NATIVE_V0_TESTNET_PREFIX
+    bech32Version = SEGWIT_V0;
+  }
+  if (bech32Prefix && bech32Version !== null) {
```
Comment (utzig) [inline src/bitcoin.js:204]: Would be better to use `bech32Prefix !== null`, no?

File: src/bitcoin.js [hunk 2]
```diff
@@ -345,25 +349,82 @@ function writeUInt64LE(n, buf, off) {
...
+
+// Determine if a a transaction should have a witness portion.
+// This will return true if any input is p2sh(p2wpkh) or p2wpkh.
+// We determine the script type based on the derivation path.
+function needsWitness(inputs) {
+  let w = false;
+  inputs.forEach((input) => {
+    if ((input.signerPath[0] === BTC_PURPOSE_P2WPKH) || 
+        (input.signerPath[0] === BTC_PURPOSE_P2SH_P2WPKH)) {
+      w = true;
```
Comment (utzig) [inline src/bitcoin.js:426]: Sorry for my JS n00b, but if it's possible to `break` a `forEach`, it could be done here...

File: src/bitcoin.js [hunk 3]
```diff
@@ -7,7 +7,11 @@ const DEFAULT_SEQUENCE = 0xffffffff;
 const DEFAULT_SIGHASH_BUFFER = Buffer.from('01', 'hex'); // SIGHASH_ALL = 0x01
 const { HARDENED_OFFSET } = require('./constants');
 const DEFAULT_CHANGE = [44 + HARDENED_OFFSET, HARDENED_OFFSET, HARDENED_OFFSET, 1, 0];
-
+const BTC_PURPOSE_P2WPKH = HARDENED_OFFSET+84;
+const BTC_PURPOSE_P2SH_P2WPKH = HARDENED_OFFSET+49;
+const BTC_LEGACY_PURPOSE = HARDENED_OFFSET+44;
+const BTC_COIN = HARDENED_OFFSET;
+const BTC_TESTNET_COIN = HARDENED_OFFSET+1;
```
Comment (alex-miller-0) [inline src/bitcoin.js:14]: These are all defined in `constants.js`. Let's use those

**Review by utzig — APPROVED** [review]
LGTM

File: src/bitcoin.js (hunk 3 above)
Comment (alex-miller-0) [reply src/bitcoin.js]: 00b5bfb

File: src/bitcoin.js (hunk 2 above)
Comment (alex-miller-0) [reply src/bitcoin.js:424]: Breaking doesn't work very well in JS. Neither does returning `true` here. There are weird rules that i've never been able to figure out

File: src/bitcoin.js (hunk 1 above)
Comment (alex-miller-0) [reply src/bitcoin.js]: 5824a88

Comment (alex-miller-0) [reply src/bitcoin.js]: 5824a88

**Pull request merged**